Source/Core/Core/HW/EXI/BBA/TAPServer.cpp

```diff
@@ -0,0 +1,305 @@
+    INFO_LOG_FMT(SP1, "Failed to set SO_NOSIGPIPE on socket\n");
+#endif
+
+  if (connect(fd, reinterpret_cast<sockaddr*>(&ss), ss_size) == -1)
+  {
+    std::string s = Common::LastStrerrorString();
+    INFO_LOG_FMT(SP1, "Couldn't connect socket ({}), unable to init BBA\n", s.c_str());
+    pi_close(fd);
+    return -1;
+  }
+
+  return fd;
+}
+
+bool CEXIETHERNET::TAPServerNetworkInterface::Activate()
+{
+  if (IsActivated())
+    return true;
+
+  m_fd = ConnectToDestination(m_destination);
+
+  INFO_LOG_FMT(SP1, "BBA initialized.");
+  return RecvInit();
+}
+
+void CEXIETHERNET::TAPServerNetworkInterface::Deactivate()
+{
+  pi_close(m_fd);
+  m_fd = -1;
+
+  m_read_enabled.Clear();
+  m_read_shutdown.Set();
+  if (m_read_thread.joinable())
+    m_read_thread.join();
+}
+
+bool CEXIETHERNET::TAPServerNetworkInterface::IsActivated()
+{
+  return (m_fd >= 0);
+}
+
+bool CEXIETHERNET::TAPServerNetworkInterface::RecvInit()
+{
+  m_read_thread = std::thread(&CEXIETHERNET::TAPServerNetworkInterface::ReadThreadHandler, this);
+  return true;
+}
+
+void CEXIETHERNET::TAPServerNetworkInterface::RecvStart()
+{
+  m_read_enabled.Set();
+}
+
+void CEXIETHERNET::TAPServerNetworkInterface::RecvStop()
+{
+  m_read_enabled.Clear();
+}
+
+bool CEXIETHERNET::TAPServerNetworkInterface::SendFrame(const u8* frame, u32 size)
+{
+  {
+    const std::string s = ArrayToString(frame, size, 0x10);
+    INFO_LOG_FMT(SP1, "SendFrame {}\n{}", size, s);
+  }
+
+  // On Windows, the data pointer is of type const char*; on other systems it is
+  // of type const void*. This is the reason for the reinterpret_cast here and
+  // in the other send/recv calls in this file.
+  u8 size_bytes[2] = {static_cast<u8>(size), static_cast<u8>(size >> 8)};
+  if (send(m_fd, reinterpret_cast<const char*>(size_bytes), 2, SEND_FLAGS) != 2)
+  {
+    ERROR_LOG_FMT(SP1, "SendFrame(): could not write size field");
+    return false;
+  }
+  int written_bytes = send(m_fd, reinterpret_cast<const char*>(frame), size, SEND_FLAGS);
+  if (u32(written_bytes) != size)
+  {
+    ERROR_LOG_FMT(SP1, "SendFrame(): expected to write {} bytes, instead wrote {}", size,
+                  written_bytes);
+    return false;
+  }
+  else
+  {
+    m_eth_ref->SendComplete();
+    return true;
+  }
+}
+
+void CEXIETHERNET::TAPServerNetworkInterface::ReadThreadHandler()
+{
+  while (!m_read_shutdown.IsSet())
+  {
+    fd_set rfds;
+    FD_ZERO(&rfds);
+    FD_SET(m_fd, &rfds);
+
+    timeval timeout;
+    timeout.tv_sec = 0;
+    timeout.tv_usec = 50000;
+    if (select(m_fd + 1, &rfds, nullptr, nullptr, &timeout) <= 0)
+      continue;
+
+    // The tapserver protocol is very simple: there is a 16-bit little-endian
+    // size field, followed by that many bytes of packet data
+    switch (m_read_state)
+    {
+    case ReadState::Size:
+    {
+      u8 size_bytes[2];
+      ws_ssize_t bytes_read = recv(m_fd, reinterpret_cast<char*>(size_bytes), 2, 0);
```

Comment:
Can't you pass a pointer to `m_read_packet_bytes_remaining` since the data is already a 16-bit little-endian integer? In the case you received just one byte, you can pass `reinterpret_cast<char*>(&m_read_packet_bytes_remaining) + 1` in `ReadState::SizeHigh`.

Reply:
Same as above - this method should be safe to build on big-endian architectures.